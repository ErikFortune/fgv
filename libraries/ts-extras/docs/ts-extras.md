<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [@fgv/ts-extras](./ts-extras.md)

## ts-extras package

## Namespaces

<table><thead><tr><th>

Namespace


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[Converters](./ts-extras.converters.md)


</td><td>


</td></tr>
<tr><td>

[Csv](./ts-extras.csv.md)


</td><td>


</td></tr>
<tr><td>

[Experimental](./ts-extras.experimental.md)


</td><td>


</td></tr>
<tr><td>

[Hash](./ts-extras.hash.md)


</td><td>


</td></tr>
<tr><td>

[RecordJar](./ts-extras.recordjar.md)


</td><td>


</td></tr>
</tbody></table>
