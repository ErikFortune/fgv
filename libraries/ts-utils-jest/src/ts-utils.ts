export * from '@fgv/ts-utils';
