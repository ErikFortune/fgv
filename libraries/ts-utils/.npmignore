# Agent-specific files
CLAUDE.md
.agents/
.claude/

# IDE and system files
.vscode/
.DS_Store

# Test and coverage files  
coverage/
*.test.ts
*.test.js
*.test.d.ts
*.test.*.map
src/test/
lib/test/
lib/**/*.test.*
lib-commonjs/test/
lib-commonjs/**/*.test.*

# Development files
*.log
.eslintrc*
jest.config.*
tsconfig.json
*.map

# Source files (if publishing only built output)
# Uncomment if you only want to publish dist/lib
src/

# Rush/Heft temporary files
.heft/
.rush/
rush-logs/
temp/

# API Extractor files (development artifacts)
etc/

# Build artifacts that shouldn't be published
dist/
lib-commonjs/

# Documentation source
docs/
*.md
!README.md
!LICENSE.md