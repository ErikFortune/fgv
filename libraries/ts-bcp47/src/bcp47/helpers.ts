/*
 * Copyright (c) 2022 Erik Fortune
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Result, mapResults, succeed } from '@fgv/ts-utils';
import { ISubtags } from './common';
import { ILanguageTagInitOptions, LanguageTag } from './languageTag';
import { LanguageSimilarityMatcher } from './match';
import { ILanguageChooserOptions, LanguageChooser } from './match/chooser';

/**
 * Any of the possible ways to represent a language - as a `string`,
 * parsed {@link Bcp47.Subtags | subtags} or an instantiated
 * {@link Bcp47.LanguageTag | language tag}.
 * @public
 */
export type LanguageSpec = string | ISubtags | LanguageTag;

/**
 * Creates a new {@link Bcp47.LanguageTag | language tag} from a {@link Bcp47.LanguageSpec | language specifier}
 *
 * The supplied initializer must be at least
 * {@link https://www.rfc-editor.org/rfc/rfc5646.html#section-2.2.9 | well-formed according to RFC 5646}.
 * Higher degrees of validation along with any normalizations may be optionally specified.
 *
 * @param from - The {@link Bcp47.LanguageSpec | language specifier} from which the tag is to
 * be created.
 * @param options - (optional) The {@link Bcp47.LanguageTagInitOptions | options} used to construct
 * and validate the tag.
 * @returns `Success` with a valid {@link Bcp47.LanguageTag | language tag} or `Failure` with details
 * if an error occurs.
 * @public
 */
// istanbul ignore next - tests applied for wrapped function
export function tag(from: LanguageSpec, options?: ILanguageTagInitOptions): Result<LanguageTag> {
  return from instanceof LanguageTag ? succeed(from) : LanguageTag.create(from, options);
}

/**
 * Creates an array of {@link Bcp47.LanguageTag | language tags} from an incoming array of
 * {@link Bcp47.LanguageSpec | language specifiers}.
 * @param from - The array of {@link Bcp47.LanguageSpec} to be converted.
 * @param options - (optional) The {@link Bcp47.LanguageTagInitOptions | options} used to construct
 * and validate any created tags.
 * @returns `Success` with an array of {@link Bcp47.LanguageTag | language tags}, or `Failure`
 * with details if an error occurs.
 * @public
 */
export function tags(from: LanguageSpec[], options?: ILanguageTagInitOptions): Result<LanguageTag[]> {
  return mapResults(from.map((f) => tag(f, options)));
}

/**
 * Determine how similar two language tags are to each other.
 *
 * @param t1 - First tag to match, supplied as one of `string`, individual
 * {@link Bcp47.Subtags | subtags}, or constructed
 * {@link Bcp47.LanguageTag | language tag}.
 * @param t2 - Second tag to match, supplied as one of `string`, individual
 * {@link Bcp47.Subtags | subtags}, or constructed
 * {@link Bcp47.LanguageTag | language tag}.
 * @param options - (optional) A set of {@link Bcp47.LanguageTagInitOptions | language tag options}
 * which control any necessary conversion or parsing.
 * @returns A numeric value in the range 1.0 (exact match) to 0.0 (no match).
 * @see For a set of common levels of similarity, see {@link Bcp47.tagSimilarity | similarity}.
 * @public
 */
// istanbul ignore next - tests applied for wrapped function
export function similarity(
  t1: LanguageSpec,
  t2: LanguageSpec,
  options?: ILanguageTagInitOptions
): Result<number> {
  return tags([t1, t2], options).onSuccess((tags) => {
    return succeed(new LanguageSimilarityMatcher().matchLanguageTags(tags[0], tags[1]));
  });
}

/**
 * Matches a list of desired {@link Bcp47.LanguageSpec | languages} to a list of available {@link Bcp47.LanguageSpec | languages},
 * return a list of matching languages ordered from best to worst.
 * @param desired - An array of {@link Bcp47.LanguageSpec | language specifications} containing an ordered list of preferred languages.
 * @param available - An array of {@link Bcp47.LanguageSpec | language specifications} containing an unordered list of available languages.
 * @param options - (optional) Parameters to control language tag conversion or comparison
 * @returns `Success` with an ordered list of matching {@link Bcp47.LanguageTag | languages}, or `Failure` with details if
 * an error occurs.
 * @public
 */
export function choose(
  desired: LanguageSpec[],
  available: LanguageSpec[],
  options?: ILanguageTagInitOptions & ILanguageChooserOptions
): Result<LanguageTag[]> {
  return tags(desired, options).onSuccess((w) => {
    return tags(available, options).onSuccess((h) => {
      return succeed(new LanguageChooser().filterLanguageTags(w, h, options));
    });
  });
}
