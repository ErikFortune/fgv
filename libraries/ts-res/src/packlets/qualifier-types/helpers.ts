/*
 * Copyright (c) 2025 Erik Fortune
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { fail, Result } from '@fgv/ts-utils';
import * as Config from './config';
import { QualifierType } from './qualifierType';
import { LanguageQualifierType } from './languageQualifierType';
import { TerritoryQualifierType } from './territoryQualifierType';
import { LiteralQualifierType } from './literalQualifierType';

/**
 * A discriminated union of all system qualifier types.
 * This allows TypeScript to properly discriminate between specific qualifier type implementations
 * and access their specific methods like getConfiguration().
 * @public
 */
export type SystemQualifierType = LanguageQualifierType | TerritoryQualifierType | LiteralQualifierType;

/**
 * Creates a {@link QualifierTypes.QualifierType | QualifierType} from a configuration object.
 * This factory function determines the appropriate qualifier type based on the systemType
 * and delegates to the appropriate type-specific createFromConfig method.
 * @param typeConfig - The {@link QualifierTypes.Config.IQualifierTypeConfig | configuration object}
 * containing the name, systemType, and optional type-specific configuration.
 * @returns `Success` with the new {@link QualifierTypes.QualifierType | QualifierType}
 * if successful, `Failure` with an error message otherwise.
 * @public
 */
export function createQualifierTypeFromConfig(
  typeConfig: Config.IAnyQualifierTypeConfig
): Result<QualifierType> {
  /* c8 ignore next 1 - defense in depth */
  const childConfig = typeConfig.configuration ?? {};
  switch (typeConfig.systemType) {
    case 'language':
      return Config.Convert.languageQualifierTypeConfig
        .convert(childConfig)
        .onSuccess((configuration) =>
          LanguageQualifierType.createFromConfig({ ...typeConfig, configuration })
        );
    case 'territory':
      return Config.Convert.territoryQualifierTypeConfig
        .convert(childConfig)
        .onSuccess((configuration) =>
          TerritoryQualifierType.createFromConfig({ ...typeConfig, configuration })
        );
    case 'literal':
      return Config.Convert.literalQualifierTypeConfig
        .convert(childConfig)
        .onSuccess((configuration) =>
          LiteralQualifierType.createFromConfig({ ...typeConfig, configuration })
        );
  }
  /* c8 ignore next 1 - defense in depth */
  return fail(`Unknown qualifier type: ${typeConfig.systemType}`);
}

/**
 * Creates a {@link QualifierTypes.SystemQualifierType | SystemQualifierType} from a system configuration object.
 * This factory function determines the appropriate qualifier type based on the systemType
 * and delegates to the appropriate type-specific createFromConfig method.
 * @param typeConfig - The {@link QualifierTypes.Config.ISystemQualifierTypeConfig | configuration object}
 * containing the name, systemType, and optional type-specific configuration.
 * @returns `Success` with the new {@link QualifierTypes.SystemQualifierType | SystemQualifierType}
 * if successful, `Failure` with an error message otherwise.
 * @public
 */
export function createQualifierTypeFromSystemConfig(
  typeConfig: Config.ISystemQualifierTypeConfig
): Result<SystemQualifierType> {
  const { systemType } = typeConfig;
  switch (systemType) {
    case 'language':
      return LanguageQualifierType.createFromConfig(typeConfig);
    case 'territory':
      return TerritoryQualifierType.createFromConfig(typeConfig);
    case 'literal':
      return LiteralQualifierType.createFromConfig(typeConfig);
  }
  /* c8 ignore next 3 - should not happen */
  // @ts-expect-error
  return fail(`${systemType}: Unknown system qualifier type.`);
}
