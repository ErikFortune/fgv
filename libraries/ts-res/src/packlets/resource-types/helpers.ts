/*
 * Copyright (c) 2025 Erik Fortune
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { fail, Result } from '@fgv/ts-utils';
import { IResourceTypeConfig } from './config';
import { ResourceType } from './resourceType';
import { JsonResourceType } from './jsonResourceType';
import { Convert as CommonConverters } from '../common';

/**
 * Creates a {@link ResourceTypes.ResourceType | ResourceType} from a configuration object.
 * @param config - The {@link ResourceTypes.Config.IResourceTypeConfig | configuration object}
 * containing the name and type name of the resource type.
 * @returns `Success` with the new {@link ResourceTypes.ResourceType | ResourceType}
 * if successful, `Failure` with an error message otherwise.
 * @public
 */
export function createResourceTypeFromConfig(config: IResourceTypeConfig): Result<ResourceType> {
  if (config.typeName === 'json') {
    return CommonConverters.resourceTypeName
      .convert(config.name)
      .onSuccess((key) => JsonResourceType.create({ key }));
  }
  return fail(`${config.typeName}: Unknown resource type.`);
}
