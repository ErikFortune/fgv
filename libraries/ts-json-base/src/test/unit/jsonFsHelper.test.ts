/*
 * Copyright (c) 2020 Erik Fortune
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import '@fgv/ts-utils-jest';

import { Converters as ExtraConverters } from '@fgv/ts-extras';
import { Converters, Validators, succeed } from '@fgv/ts-utils';
import { IMockFileConfig, MockFileSystem } from '@fgv/ts-utils-jest/lib/helpers/fsHelpers';
import fs from 'fs';
import { JsonFile, JsonValue } from '../..';

describe('JsonFsHelper class', () => {
  const mockDate = new Date();
  const mockGoodPath = 'path/to/some/file.json';
  const mockGoodPayload = {
    someProperty: 'some value',
    prop: [1, 2],
    now: mockDate.toISOString()
  };

  const mockBadPath = 'path/to/some/malformed.json';
  const mockFsConfig: IMockFileConfig[] = [
    {
      path: mockGoodPath,
      payload: JSON.stringify(mockGoodPayload),
      writable: false
    },
    {
      path: mockBadPath,
      payload: '{ bad json',
      writable: false
    }
  ];

  let helper: JsonFile.JsonFsHelper;

  beforeEach(() => {
    helper = JsonFile.DefaultJsonFsHelper;
  });

  describe('readJsonFileSync function', () => {
    test('returns a requested json file', () => {
      const mockFs = new MockFileSystem(mockFsConfig);
      const spies = mockFs.startSpies();
      expect(helper.readJsonFileSync(mockGoodPath)).toSucceedWith(mockGoodPayload);
      expect(spies.read).toHaveBeenCalledTimes(1);
      spies.restore();
    });

    test('fails for malformed json', () => {
      const mockFs = new MockFileSystem(mockFsConfig);
      const spies = mockFs.startSpies();
      expect(helper.readJsonFileSync(mockBadPath)).toFailWith(/in JSON at position/i);
      expect(spies.read).toHaveBeenCalledTimes(1);
      spies.restore();
    });

    test('propagates any error', () => {
      const path = 'path/to/some/file.json';
      jest.spyOn(fs, 'readFileSync').mockImplementation((gotPath: unknown) => {
        if (typeof gotPath !== 'string') {
          throw new Error('Mock implementation only accepts string');
        }
        expect(gotPath).toContain(path);
        throw new Error('Mock Error!');
      });

      expect(helper.readJsonFileSync(path)).toFailWith(/mock error/i);
    });
  });

  describe('processJsonFileSync function', () => {
    test('converts a well-formed JSON file', () => {
      const mockConverter = Converters.object({
        someProperty: Converters.string,
        prop: Converters.arrayOf(Converters.number),
        now: ExtraConverters.isoDate
      });
      const mockConverted = {
        ...mockGoodPayload,
        now: mockDate
      };

      const mockFs = new MockFileSystem(mockFsConfig);
      const spies = mockFs.startSpies();
      spies.clear();
      expect(helper.convertJsonFileSync(mockGoodPath, mockConverter)).toSucceedWith(mockConverted);
      expect(spies.read).toHaveBeenCalledTimes(1);
      spies.restore();
    });

    test('validates a well-formed JSON file', () => {
      const mockValidator = Validators.object({
        someProperty: Validators.string,
        prop: Validators.arrayOf(Validators.number),
        now: Validators.string
      });
      const mockValidated = {
        ...mockGoodPayload,
        now: mockDate.toISOString()
      };

      const mockFs = new MockFileSystem(mockFsConfig);
      const spies = mockFs.startSpies();
      spies.clear();
      expect(helper.convertJsonFileSync(mockGoodPath, mockValidator)).toSucceedWith(mockValidated);
      expect(spies.read).toHaveBeenCalledTimes(1);
      spies.restore();
    });
  });

  describe('processJsonDirectorySync function', () => {
    interface IThing {
      name: string;
      optionalString?: string;
    }
    describe('with a converter', () => {
      const thing = Converters.object<IThing>(
        {
          name: Converters.string,
          optionalString: Converters.string
        },
        {
          optionalFields: ['optionalString']
        }
      );
      const options = { converter: thing };

      test('reads JSON files from a folder, ignoring non-JSON', () => {
        expect(
          helper.convertJsonDirectorySync('src/test/unit/data/file/good', options)
        ).toSucceedAndMatchSnapshot();
      });

      test('fails for a non-folder', () => {
        expect(
          helper.convertJsonDirectorySync('src/test/unit/data/file/good/thing1.json', options)
        ).toFailWith(/not a directory/i);
      });

      test('fails by default if any of the items in the folder fail conversion', () => {
        expect(helper.convertJsonDirectorySync('src/test/unit/data/file/bad', options)).toFailWith(
          /bad3.json/i
        );
      });
    });

    describe('with a validator', () => {
      const thing = Validators.object<IThing>(
        {
          name: Validators.string,
          optionalString: Validators.string
        },
        {
          options: {
            optionalFields: ['optionalString']
          }
        }
      );
      const options = { converter: thing };

      test('reads JSON files from a folder, ignoring non-JSON by default', () => {
        expect(helper.convertJsonDirectorySync('src/test/unit/data/file/good', options)).toSucceedAndSatisfy(
          (items) => {
            expect(
              (items as Array<{ filename: string }>).every((item) => item.filename.endsWith('.json'))
            ).toBe(true);
            expect(items).toMatchSnapshot();
          }
        );
      });

      test('reads JSON files from a folder, using file filter if supplied', () => {
        const options2 = { ...options, files: [/.*.txt/] };
        expect(helper.convertJsonDirectorySync('src/test/unit/data/file/good', options2)).toSucceedAndSatisfy(
          (items) => {
            expect(
              (items as Array<{ filename: string }>).every((item) => item.filename.endsWith('.txt'))
            ).toBe(true);
            expect(items).toMatchSnapshot();
          }
        );
      });
    });
  });

  describe('processJsonDirectoryToMapSync function', () => {
    interface IThing {
      name: string;
      optionalString?: string;
    }
    const thing = Converters.object<IThing>({
      name: Converters.string,
      optionalString: Converters.string.optional()
    });
    const options = { converter: thing };

    test('reads JSON files from a folder, ignoring non-JSON by default', () => {
      expect(
        helper.convertJsonDirectoryToMapSync('src/test/unit/data/file/good', options)
      ).toSucceedAndMatchSnapshot();
    });

    test('applies a name transformation to the returned map if supplied', () => {
      const transformName: JsonFile.ItemNameTransformFunction<IThing> = (n, t) => {
        expect(t).toBeDefined();
        return succeed(`thing:${n}`);
      };
      const myOptions = { ...options, transformName };
      expect(
        helper.convertJsonDirectoryToMapSync('src/test/unit/data/file/good', myOptions)
      ).toSucceedAndMatchSnapshot();
    });

    test('fails for a non-folder', () => {
      expect(
        helper.convertJsonDirectoryToMapSync('src/test/unit/data/file/good/thing1.json', options)
      ).toFailWith(/not a directory/i);
    });

    test('fails by default if any of the items in the folder fail conversion', () => {
      expect(helper.convertJsonDirectoryToMapSync('src/test/unit/data/file/bad', options)).toFailWith(
        /bad3.json/i
      );
    });
  });

  describe('writeJsonFileSync function', () => {
    test('saves to the requested json file', () => {
      const path = 'path/to/some/file.json';
      const payload = { someProperty: 'some value', prop: [1, 2] };
      const spy = jest
        .spyOn(fs, 'writeFileSync')
        .mockImplementation((gotPath: unknown, gotPayload: unknown) => {
          if (typeof gotPath !== 'string' || typeof gotPayload !== 'string') {
            throw new Error('Mock implementation only accepts string');
          }
          expect(gotPath).toContain(path);
          expect(JSON.parse(gotPayload)).toEqual(payload);
        });

      expect(helper.writeJsonFileSync(path, payload)).toSucceedWith(true);
      spy.mockRestore();
    });

    test('fails for undefined by default', () => {
      const path = 'path/to/some/file.json';
      const payload = undefined;
      const spy = jest
        .spyOn(fs, 'writeFileSync')
        .mockImplementation((gotPath: unknown, gotPayload: unknown) => {
          if (typeof gotPath !== 'string') {
            throw new Error('Mock implementation only accepts string');
          }
          expect(gotPath).toContain(path);
          expect(gotPayload).toBeUndefined();
        });

      expect(helper.writeJsonFileSync(path, payload as unknown as JsonValue)).toFailWith(
        /could not stringify undefined/i
      );
      spy.mockRestore();
    });

    test('saves undefined if configured', () => {
      helper = new JsonFile.JsonFsHelper({ allowUndefinedWrite: true });
      const path = 'path/to/some/file.json';
      const payload = undefined;
      const spy = jest
        .spyOn(fs, 'writeFileSync')
        .mockImplementation((gotPath: unknown, gotPayload: unknown) => {
          if (typeof gotPath !== 'string') {
            throw new Error('Mock implementation only accepts string');
          }
          expect(gotPath).toContain(path);
          expect(gotPayload).toBeUndefined();
        });

      expect(helper.writeJsonFileSync(path, payload as unknown as JsonValue)).toSucceedWith(true);
      spy.mockRestore();
    });

    test('propagates an error', () => {
      const path = 'path/to/some/file.json';
      const payload = { someProperty: 'some value', prop: [1, 2] };
      const spy = jest.spyOn(fs, 'writeFileSync').mockImplementation((gotPath: unknown) => {
        if (typeof gotPath !== 'string') {
          throw new Error('Mock implementation only accepts string');
        }
        expect(gotPath).toContain(path);
        throw new Error('Mock Error!');
      });

      expect(helper.writeJsonFileSync(path, payload)).toFailWith(/mock error/i);
      spy.mockRestore();
    });
  });
});
