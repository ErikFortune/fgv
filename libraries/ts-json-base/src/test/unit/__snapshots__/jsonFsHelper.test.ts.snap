// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonFsHelper class processJsonDirectorySync function with a converter reads JSON files from a folder, ignoring non-JSON: toSucceedAndMatchSnapshot 1`] = `
Array [
  Object {
    "filename": "thing1.json",
    "item": Object {
      "name": "thing1",
      "optionalString": "thing 1 optional string",
    },
  },
  Object {
    "filename": "thing2.json",
    "item": Object {
      "name": "thing2",
    },
  },
]
`;

exports[`JsonFsHelper class processJsonDirectorySync function with a validator reads JSON files from a folder, ignoring non-JSON by default 1`] = `
Array [
  Object {
    "filename": "thing1.json",
    "item": Object {
      "name": "thing1",
      "optionalString": "thing 1 optional string",
    },
  },
  Object {
    "filename": "thing2.json",
    "item": Object {
      "name": "thing2",
    },
  },
]
`;

exports[`JsonFsHelper class processJsonDirectorySync function with a validator reads JSON files from a folder, using file filter if supplied 1`] = `
Array [
  Object {
    "filename": "thing9.txt",
    "item": Object {
      "name": "thing2",
    },
  },
]
`;

exports[`JsonFsHelper class processJsonDirectoryToMapSync function applies a name transformation to the returned map if supplied: toSucceedAndMatchSnapshot 1`] = `
Map {
  "thing:thing1" => Object {
    "name": "thing1",
    "optionalString": "thing 1 optional string",
  },
  "thing:thing2" => Object {
    "name": "thing2",
  },
}
`;

exports[`JsonFsHelper class processJsonDirectoryToMapSync function reads JSON files from a folder, ignoring non-JSON by default: toSucceedAndMatchSnapshot 1`] = `
Map {
  "thing1" => Object {
    "name": "thing1",
    "optionalString": "thing 1 optional string",
  },
  "thing2" => Object {
    "name": "thing2",
  },
}
`;
